markdown.dtx

```diff
@@ -19167,6 +19167,78 @@ end
 %    \end{macrocode}
 % \par
 % \begin{markdown}
+% The most comfortable way to to parse the jekyllData is using \pkg{expl3}'s
+% l3keys module. due to the structure of the data we need to initialize some
+% variables to save the data.
+% \end{markdown}
+%  \begin{macrocode}
+\ExplSyntaxOn
+\clist_new:N \l__markdown_jekyll_data_clist
+\seq_new:N \l__markdown_jekyll_data_seq
+\prop_new:N \l__markdown_jekyll_prop
+%    \end{macrocode}
+% \par
+% \begin{markdown}
+% The default setup will add unknown data to a property list
+% \m{l__markdown_jekyll_prop} and set \m{title}, \m{author} and \m{date} to
+% be used with \m{maketitle}.
+% \end{markdown}
+%    \begin{macrocode}
+\keys_define:nn {markdown/JekyllData} {
+  title .code:n = \title{#1}
+    \prop_put:NVn \l__markdown_jekyll_prop \l_keys_key_tl {#1},
+  author .code:n = \author{#1}
+    \prop_put:NVn \l__markdown_jekyll_prop \l_keys_key_tl {#1},
+  date .code:n = \date{#1}
+    \prop_put:NVn \l__markdown_jekyll_prop \l_keys_key_tl {#1},
+  maketitle .bool_set:N = \l__markdown_maketitle_bool,
+  maketitle .initial:n = true,
+  unknown .code:n = {
+    \prop_put:NVn \l__markdown_jekyll_prop \l_keys_key_tl {#1}
+  },
+}
+%    \end{macrocode}
+% \par
+% \begin{markdown}
+% The `unknown` value allows the user to add arbitrary keys to the jekyll
+% Syntax Internally sequences will be saved with l3seq but the default
+% setup will convert the output to a nested comma list which will be
+% saved in the propertylist to be accessed by the user.
+% \end{markdown}
+%    \begin{macrocode}
+\markdownSetup{rendererPrototypes={
+  jekyllDataEnd = \bool_if:NT \l__markdown_maketitle_bool \maketitle,
+  jekyllDataString={
+    \tl_if_empty:NTF \l__markdown_jekyll_sequence_tl
+      {\keys_set:nn {markdown/JekyllData}{#1 = {#2}}}
+      {\clist_put_right:Nn \l__markdown_jekyll_data_clist {{#1} = {#2}}}
+  },
+  jekyllDataSequenceBegin={
+    \tl_set:Nn \l__markdown_jekyll_sequence_tl  {#1}
+    \clist_clear:N \l__markdown_jekyll_data_clist
+  },
+  jekyllDataSequenceEnd={
+    \seq_if_empty:NTF \l__markdown_jekyll_data_seq {
+      \exp_args:Nnx \keys_set:nn {markdown/JekyllData}{
+        {\l__markdown_jekyll_sequence_tl} =
```

Comment:
Do we handle nesting? I would assume that for the following document, we should end up with the command `\keys_set:nn {markdown/JekyllData} {author / author / author = John Doe}`. However, `\l__markdown_jekyll_sequence_tl` is always overriden in `jekyllDataSequenceBegin`, not appended, and we seem to completely ignore `jekyllDataMappingBegin`.
```
author:
  author:
    author: John Doe
```

Reply:
currently we don't handle nesting, but that's one of the questions I wanted to ask. 

The suggestion you made does not work, because you wanted it to be a sequence - then we should rename that one because it's not a sequence anymore.

Reply:
I also thought about what would be a good structure for nesting because we have to save it in datatypes. If we shall support nesting i'd save the key inside a sequence used as a stack  so on could always reference the most inner key. So it's a big deal to add it.

Reply:
Perhaps we could keep the current address in a clist and then concatenate with the `/` delimiter when we are calling `\keys_set:nn {markdown/JekyllData}`? In a pseudocode, something like this:
``` tex
\clist_new:N \l_my_clist
\markdownSetup{
  rendererPrototypes = {
    jekyllDataMappingBegin = {
      \clist_push_right:Nn \l_my_clist {{#1}}
    },
    jekyllDataSequenceBegin = {
      \clist_push_right:Nn \l_my_clist {{#1}}
    },
    jekyllDataString = {
      \keys_set:nn {markdown/jekyllData} { \seq_use:Nn \l_tmpa_seq { / } / {#1} = {#2}}
    },
    jekyllDataMappingEnd = {
      \clist_pop_right:N \l_my_clist
    },
    jekyllDataSequenceEnd = {
      \clist_pop_right:N \l_my_clist
    },
  }
}
```

Reply:
For example, assume we have the following example document:
``` yaml
title: Document title
author:
- name: Author One
- name: Author Two
```
Then, the expl code would call the following commands (not necessarily in this order):
``` tex
\keys_set:nn {markdown/jekyllData} {title} {Document title}
\keys_set:nn {markdown/jekyllData} {author/1/name} {Author One}
\keys_set:nn {markdown/jekyllData} {author/2/name} {Author Two}
```
You may wonder whether keeping the sequence indices is useful, since it makes difficult to define how the individual keys should be handled. Therefore, we could instead adopt the following notation:
``` tex
\keys_set:nn {markdown/jekyllData} {title} {Document title}
\keys_set:nn {markdown/jekyllData} {author/N/name} {Author One}
\keys_set:nn {markdown/jekyllData} {author/N/affiliation} {University of Somewhere}
\keys_set:nn {markdown/jekyllData} {author/N/name} {Author Two}
\keys_set:nn {markdown/jekyllData} {author/N/affiliation} {University of Nowhere}
```
We can achieve this by altering the above pseudocode a little:
``` diff
diff --git a/pseudocode b/pseudocode
index 49a8c71..faf8dc0 100644
--- a/pseudocode
+++ b/pseudocode
@@ -5,7 +5,7 @@
       push {#1} to \l_my_clist
     },
     jekyllDataSequenceBegin = {
-      push {#1} to \l_my_clist
+      push N to \l_my_clist
     },
     jekyllDataString = {
       \keys_set:nn {markdown/jekyllData} { items of \l_my_clist interleaved with slashes / {#1} = {#2} }
```
This already allows the users to make meaningful definitions, such as this one (again, pseudocode):
``` latex
\keys_define:nn {markdown/JekyllData} {
  title .code:n = \title{#1}
  author/N/name .code:n = {add {#1} to a list of author names},
}
\markdownSetup{
  jekyllDataEnd = {
    \author{all author names interleaved with \and}
    \maketitle
  }
}
```

Reply:
That's exactly what I thought about when I said nesting is not a big issue. I'd just prefer if we'd rename the sequence then. Because the name would make people think that the order matters.

Reply:
Oh, absolutely. Let's first get the implementation right, then we can quibble about names if needed.

Reply:
It's just a question whether we want to create:

1. an event-based system, where users register callbacks via l3keys (see above),

2. a data structure that the users can walk (your current code), I assume? I still find it difficult to read expl3 code.

Reply:
We have one issue: 

```
  author/N/name .code:n = {add {#1} to a list of author names},
  author/N/affiliation .code:n = {add {#1} to a list of author affiliations},
```
would lead to a conflict in case one author has an affiliation and one doesn't.

Alternative would be to create a property list and set the value of author to some fixed value indicatng that one. 
Or I create a keyval structure as I did before and author gets the keyval list assigned.

Reply:
May mappings also be nested? because  then we'd have to do it different…

Reply:
> 1.  an event-based system, where users register callbacks via l3keys (see above),
> 2. a data structure that the users can walk (your current code I assume? I still find it difficult to read expl3 code).

I replaced the last commit. I now have allowed nesting and the default handler for unknown keys will save them to a property list where the key is the module used to save.

Users can create handles for their keys themselves, like I did for `\title` the unknown handler will only be used if no alternative is found.

Then the user can access the value from the property list. In expl3 this would be 
```
\prop_item:Nn \l__markdown_jekyll_prop {author/N/affiliation} 
```
We can add a l2 macro to access it which used the key as argument.

Reply:
> May mappings also be nested? because then we'd have to do it different…

They can:

```yaml
author:
  name: John
  surname: Doe
```

I think that you can easily extend your current code to treat sequences and mappings the same way:

``` latex
    jekyllDataMappingBegin={
      \seq_put_right:Nn \l__markdown_jekyll_data_seq {#1}
      \exp_args:Nx \keys_define:nn
        {markdown/JekyllData/\seq_use:Nn \l__markdown_jekyll_data_seq {/}}
        {unknown .code:n = {
          \exp_args:Nnx \keys_set:nn
           {markdown/JekyllData}{\l_keys_key_tl  = \exp_not:n {##1}}}}
    },
    jekyllDataMappingEnd={
      \seq_pop_right:NN \l__markdown_jekyll_data_seq \l_tmpa_tl
    },
```

Reply:
hmmm i could use the number as a key. So the sequence would not have a key/item but i use the ID.

Reply:
> would lead to a conflict in case one author has an affiliation and one doesn't.

You're right, of course. But it's not clear how to mitigate this. If we call `\keys_set:nn` as soon as we encounter `author/N/name`, then we don't have any guarantee that we have already seen `author/N/affiliation`. We could solve this by separating the `\prop_put:NVn` and `\keys:set:nn`:

1. In `jekyllDataString`, we always call `\prop_put:NVn` to set `markdown/jekyllData/.../KEY` property.
2. In `jekyllDataEnd`, we walk all properties in `\l__markdown_jekyll_prop` and call `\keys:set:nn`.

This way, all properties are already set when `\keys:set:nn` is called, so for example the user can query whether there exists an `author/N/affiliation` from the l3keys callback for `author/N/name`. This allows for a nice combination of event-based and random-access processing.

Reply:
this would only work if we don't have multiple levels nested. 

I will use the Number so `author/N/property` if there's a sequence inside. Then a sequence and mapping would indeed be treated the same way.  We could then add that acessing the value author automatically iterates but i'd not add this generally because the possibilities are too many.

Reply:
> We could then add that acessing the value author automatically iterates but i'd not add this generally because the possibilities are too many.

We could call `\prop_put:NVn` with `author/1/name`, but `\keys:set:nn` with `author/N/name` (a literal `N`) and store the `N` in a variable. This way, users could easily define l3keys code that would react to any `author/N/name` without having to spell out `author/1/name`, `author/2/name`, ..., but they would know the value of `N` and could lookup the value of `author/N/affiliation` for that specific `N`.

Reply:
> this would only work if we don't have multiple levels nested.

How so? It seems to me that we can easily do this with the current data types. The only trouble is that I suck at expl3, but the algorithm seems quite straightforward.

Reply:
I will have to check this. Since we already have the amount of elements inside the macros it's not a big deal to add it with a literal `N` for a default setup this should work. But it will probably not with users extending it. So I'd probably only store number of elements to  `author/N` and users could map it depending on their datatypes.  For authors this is fine but it's getting more complicated with multiple levels nested. 

I have to continue my work for today but can implement this in the evening.

Reply:
However, I am learning a lot from your code and from skimming over `interface3.pdf`, so I think I can try to cook something up if you're getting tired of it.

Reply:
Oh, but what if we have multiple nesting levels, right? What if we were processing `authors/N/N/name`. Then `\l__markdown_jekyll_N` would need to be a sequence. Ugh!

Reply:
We could have e.g. the `#authors` property, which would hold the total number of properties in `authors`. However, I don't think that we necessarily need this in a minimal implementation.

Reply:
i just added `authors/N` for this i think it's easier to store them in the same module.